### ISSUE TITLE
Deadlock when calling supervise()

### ISSUE THREAD

**Issue by ederst** [jenkinsci/mesos-plugin#97]
It seems when `JenkinsScheduler.statusUpdate()` tries to stop the Scheduler and the Retention Timer of a Slave tries to stop a Slave it can somehow end in a deadlock.

This is because the Timer locks the `MesosImpl` instance and `statusUpdate()` the `SUPERVISOR_LOCK`. Then `MesosImpl` tries to terminate the Slave and waits for the `SUPERVISOR_LOCK` to be freed by the `statusUpdate()` Thread. However, it seems that `statusUpdate()` needs a lock on `MesosImpl` too, when trying to stop the Scheduler.

This is the Threaddump (I use a slightly modified version of Mesos plugin 0.6.0, so the linenumbers are probably not 100% right):
```
"Thread-2516073" - Thread t@2898790
   java.lang.Thread.State: BLOCKED
    at org.jenkinsci.plugins.mesos.Mesos$MesosImpl.stopScheduler(Mesos.java:141)
    - waiting to lock <62132b60> (a org.jenkinsci.plugins.mesos.Mesos$MesosImpl) owned by "jenkins.util.Timer [#9]" t@66
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.supervise(JenkinsScheduler.java:749)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.statusUpdate(JenkinsScheduler.java:634)

   Locked ownable synchronizers:
    - locked <3af5466a> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
```

```
"jenkins.util.Timer [#9]" - Thread t@66
   java.lang.Thread.State: WAITING
    at sun.misc.Unsafe.park(Native Method)
    - waiting to lock <3af5466a> (a java.util.concurrent.locks.ReentrantLock$NonfairSync) owned by "Thread-2516073" t@2898790
    at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:834)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(AbstractQueuedSynchronizer.java:867)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(AbstractQueuedSynchronizer.java:1197)
    at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(ReentrantLock.java:214)
    at java.util.concurrent.locks.ReentrantLock.lock(ReentrantLock.java:290)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.supervise(JenkinsScheduler.java:725)
    at org.jenkinsci.plugins.mesos.JenkinsScheduler.terminateJenkinsSlave(JenkinsScheduler.java:220)
    - locked <55398768> (a org.jenkinsci.plugins.mesos.JenkinsScheduler)
    at org.jenkinsci.plugins.mesos.Mesos$MesosImpl.stopJenkinsSlave(Mesos.java:157)
    - locked <62132b60> (a org.jenkinsci.plugins.mesos.Mesos$MesosImpl)
    at org.jenkinsci.plugins.mesos.MesosComputerLauncher.terminate(MesosComputerLauncher.java:122)
    at org.jenkinsci.plugins.mesos.MesosSlave.terminate(MesosSlave.java:91)
    at org.jenkinsci.plugins.mesos.MesosRetentionStrategy.check(MesosRetentionStrategy.java:70)
    - locked <75b63404> (a org.jenkinsci.plugins.mesos.MesosRetentionStrategy)
    at org.jenkinsci.plugins.mesos.MesosRetentionStrategy.check(MesosRetentionStrategy.java:26)
    at hudson.slaves.ComputerRetentionWork.doRun(ComputerRetentionWork.java:66)
    at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:54)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask$Sync.innerRunAndReset(FutureTask.java:351)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:722)

   Locked ownable synchronizers:
    - locked <703c7665> (a java.util.concurrent.ThreadPoolExecutor$Worker)
```

I tried to solve the problem myself, but I somehow got a knot in my brain from all the synchronized calls etc. The only thing I can guess is that the multiple synchronized cross calls between `MesosImpl` and `JenkinsScheduler` are not great.

Maybe some Java whiz can solve the problem there.

**Comment by ederst** [comment]
Also posted this here on Jenkin Jira: https://issues.jenkins-ci.org/browse/JENKINS-27670

**Comment by ederst** [comment]
Some news on that issue.

The lock seems to occur because Jenkins terminates and removes the Slave via the Retention check (`Jenkins.getInstance().removeNode(this)`). This causes the docker container on the Mesos node to exit and Mesos sends a status update for the underlying Task to Jenkins that it has finished. Consqeuently, the two threads will eventually battle for the the `supervise()` method.

The main issue I have with this is that at first Jenkins terminates the Slave and then notifies Mesos to kill the underlying Task. However, the Task was already finished by Mesos (or by the fact that the Docker container exited).

Wouldn't it make more sense that the Retention check notifies the Mesos driver that the Task should be killed/finished and then let `statusUpdate()` remove the Slave from the Jenkins instance?

At least this would prevent the race for the `supervise()` method and also would prevent that Jenkins sends a `driver.kill(taskId)` for a Taks which is already killed/finished.

**Comment by ederst** [comment]
Also (this is just a hypothesis): Due to the handling of killing the Tasks and Jenkins slaves this probably leads to some Tasks in the Framework being reported as "KILLED" and some being reported as "FINISHED". Might be that sometimes the `Result` for the `taskId` is not removed by `statusUpdate()`  before `JenkinsScheduler.terminateJenkinsSlave()` calls `driver.kill()`...

**Comment by maselvaraj** [comment]
@ederst  I believe your PR https://github.com/jenkinsci/mesos-plugin/pull/101 addresses this issue. Have you tested this fix on your environment ?

**Comment by ederst** [comment]
Completely forgot about this issue. But I'm not sure if #101 will fix this, because the problem here is the cross method calling between `MesosImpl` and `JenkinsScheduler`.

I think we tried to fix this by moving method `supervise()` to `MesosImpl`.

Have to test it to be sure, though.

**Comment by adaughterson** [comment]
My tasks are always Killed, with an stack dump that looks like this:

INFO: Disconnecting idle computer mesos-jenkins-a07bc491-1b90-408c-a919-54f45cc69708
Jun 01, 2015 8:23:11 AM org.jenkinsci.plugins.mesos.MesosSlave terminate
INFO: Terminating slave mesos-jenkins-a07bc491-1b90-408c-a919-54f45cc69708
Jun 01, 2015 8:23:11 AM org.jenkinsci.plugins.mesos.JenkinsScheduler terminateJenkinsSlave
INFO: Terminating jenkins slave mesos-jenkins-a07bc491-1b90-408c-a919-54f45cc69708
Jun 01, 2015 8:23:11 AM org.jenkinsci.plugins.mesos.JenkinsScheduler terminateJenkinsSlave
INFO: Killing mesos task value: "mesos-jenkins-a07bc491-1b90-408c-a919-54f45cc69708"

Jun 01, 2015 8:23:11 AM hudson.remoting.AbstractByteArrayCommandTransport$1 handle
WARNING: Failed to construct Command
java.io.EOFException
        at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2328)
        at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:2797)
        at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:802)
        at java.io.ObjectInputStream.<init>(ObjectInputStream.java:299)
        at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:40)
        at hudson.remoting.AbstractByteArrayCommandTransport$1.handle(AbstractByteArrayCommandTransport.java:61)
        at org.jenkinsci.remoting.nio.NioChannelHub$2.run(NioChannelHub.java:594)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
        at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

Could these be symptoms of this issue?

**Comment by carlossg** [comment]
I think this is fixed with #196 and can be closed
